050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,821 @@
+    * If any gap will exist in the existing node IDs, it will be filled in.
+      For example when the existing nodes have IDs 0, 1, 4, 5, 6; the next node will get ID 2.
+* By default, when scaling down a pool, a highest used node ID will be remove.
+    * For example in Kafka cluster with two pool, where the first pool has IDs 0, 1 and 2 and the second pool has IDs 3, 4 and 5 ... when you scale down the first pool, the node with ID 2 will be removed and when you scale down the second pool node 5 will be removed.
+* The pods will reflect the node IDs in their name in the same way as they do today.
+
+The fact that the node IDs will not be assigned to the pools means the Pods will be mixed.
+In a cluster with two pools, they might look like this:
+
+* `my-cluster-big-nodes-0`
+* `my-cluster-big-nodes-1`
+* `my-cluster-small-nodes-2`
+* `my-cluster-small-nodes-3`
+* `my-cluster-big-nodes-4`
+* `my-cluster-small-nodes-5`
+
+This is anyway unavoidable if we want to allow nodes to move between the pools.
+
+To allow the operator to keep track of the assigned node IDs, it will keep track of them in the `.status` section of the `KafkaNodePool` resource.
+That way, it will be able to know at any point in time which IDs are used and which are available.
+It will be also able to identify the scale-down or scale-up events using this information.
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: big-nodes
+  # ...
+spec:
+  replicas: 3
+  # ...
+status:
+  nodeIds:
+    - 0
+    - 1
+    - 2
+  clusterId: cABpUKeDTqi7q_XOtBMgUv
+  # replicas and labelSelector fields are needed for the scale subresource
+  replicas: 3
+  labelSelector: >-
+    strimzi.io/cluster=my-cluster,strimzi.io/name=m-cluster-big-nodes,strimzi.io/kind=Kafka
+```
+
+User will be able to use annotations to indicate to the operator what should be the next ID when scaling up or what should be the node ID which should be removed when scaling down through annotations.
+Setting an annotation `strimzi.io/next-node-ids` will tell the operator what should be the next ID used in scale-up.
+The value might be a single number, list of numbers or one or more ranges.
+E.g. `[3]`, `[3, 4, 5]` or `[1000-1010]`.
+Similarly, setting an annotation `strimzi.io/remove-node-ids` will allow to configure IDs of node which should be removed.
+This will support only an ordered list of nodes IDs without ranges.
+When the operator is adding or removing a node from the pool, it will look at these annotations and if they are set, it will use them to override the default mechanism for picking up the next node ID or the node to remove.
+These annotations will be ignored when scaling is not requested by changing `.spec.replicas`.
+If these annotations cannot be fulfilled (for example because the ID is already taken and already in use) they will be ignored and the next suitable ID will be picked up.
+The operator will raise a warning in its log in such case.
```

Comment:
1. I wonder if we should list exactly the cases which we'll cover. 
2. Is it enough just to log a warning?
3. I assume it's not allowed for these things to overlap between pools?

Reply:
> 1. I wonder if we should list exactly the cases which we'll cover.

Sorry, what exactly you mean here? what would be the cases?

> 2. Is it enough just to log a warning?

TBH, none of this is implemented so far. The prototype just picks the next ID right now. But my plan was to treat these as an indication and not fail when they don't fit. I think failing with error is a bit dangerous if you combine it with some auto-scaling mechanisms where the automated scale-up could possibly lead to reconciliation errors. But I guess its _open for discussion_.

> 3. I assume it's not allowed for these things to overlap between pools?

I do not see overlap as an issue. I can for example imagine that you would want to have your controller nodes from the 0-99 range and all your broker pools from 100-1000 range. I don't think the overlap causes any issues - since the node IDs are picked serially, it will simply always get the next free one.